File: android/app/src/main/java/com/hedvig/app/AsyncStorageNativeReaderImpl.java
package com.hedvig.app;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.facebook.react.modules.storage.ReactDatabaseSupplier;
import com.hedvig.android.owldroid.util.react.AsyncStorageNativeReader;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

public class AsyncStorageNativeReaderImpl implements AsyncStorageNativeReader {

    private final Context context;

    @Inject
    public AsyncStorageNativeReaderImpl(Context context) {
        this.context = context;
    }

    @NotNull
    @Override
    public String getKey(@NotNull String key) {
        SQLiteDatabase readableDatabase = null;
        Cursor catalystLocalStorage = null;
        try {
            readableDatabase = ReactDatabaseSupplier.getInstance(context).getReadableDatabase();
            catalystLocalStorage = readableDatabase.query(
                    "catalystLocalStorage",
                    new String[]{"key", "value"},
                    "key = ?",
                    new String[]{key},
                    null,
                    null,
                    null
            );
            if (catalystLocalStorage.moveToFirst()) {
                return catalystLocalStorage.getString(catalystLocalStorage.getColumnIndex("value"));
            } else {
                throw new RuntimeException(String.format("Could not find key %s", key));
            }
        } finally {
            if (catalystLocalStorage != null) {
                catalystLocalStorage.close();
            }
            if (readableDatabase != null) {
                readableDatabase.close();
            }
        }
    }
}

File: android/app/src/main/java/com/hedvig/app/AsyncStorageNativeReaderImpl.kt
package com.hedvig.app

import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase

import com.facebook.react.modules.storage.ReactDatabaseSupplier
import com.hedvig.android.owldroid.util.react.AsyncStorageNativeReader

import javax.inject.Inject

class AsyncStorageNativeReaderImpl @Inject
constructor(private val context: Context) : AsyncStorageNativeReader {

    override fun getKey(key: String): String {
        var readableDatabase: SQLiteDatabase? =
            null
        var catalystLocalStorage: Cursor? = null
        try {
            readableDatabase = ReactDatabaseSupplier.getInstance(context).readableDatabase
            catalystLocalStorage = readableDatabase.query(
                "catalystLocalStorage",
                arrayOf("key", "value"),
                "key = ?",
                arrayOf(key), null, null, null
            )

            return if (catalystLocalStorage!!.moveToFirst()) {
                catalystLocalStorage.getString(catalystLocalStorage.getColumnIndex("value"))
            } else {
                throw RuntimeException(String.format("Could not find key %s", key))
            }
        } finally {
            catalystLocalStorage?.close()
            readableDatabase?.close()
        }
    }

    override fun deleteKey(key: String) {
        var readableDatabase: SQLiteDatabase? =
            null
        try {
            readableDatabase = ReactDatabaseSupplier.getInstance(context).readableDatabase
            readableDatabase?.delete("catalystLocalStorage", "key = ?", arrayOf(key))
        } finally {
            readableDatabase?.close()
        }
    }
}


